pkg/medusa/hostmap.go

```diff
@@ -0,0 +1,206 @@
+			location := NodeLocation{
+				Rack: i.Rack,
+				DC:   i.Datacenter,
+			}
+			IP := HostDNSOrIP(i.Host)
+			if err != nil {
+				return nil, err
+			}
+			_, exists := out[location]
+			if exists {
+				out[location] = append(out[location], IP)
+			} else {
+				out[location] = []HostDNSOrIP{IP}
+			}
+		}
+	}
+	return out, nil
+}
+
+// getClusterRackFQDNs gets a map of racks to FQDNs from the current K8ssandraCluster k8s object. The CassDC does not exist yet, so we cannot refer to it for names.
+// We refer to the following code for how to calculate pod names: // https://github.com/k8ssandra/cass-operator/blob/master/pkg/reconciliation/construct_statefulset.go#L39
+func getTargetRackFQDNs(Kluster k8ssandraapi.K8ssandraCluster, dcName string) (map[NodeLocation][]HostName, error) {
+	cassDC, err := cassDCFromKluster(Kluster, dcName)
+	if err != nil {
+		return nil, err
+	}
+	racks := cassDC.GetRacks()
+	out := make(map[NodeLocation][]HostName)
+	for _, i := range racks {
+		location := NodeLocation{
+			DC:   cassDC.Name,
+			Rack: i.Name,
+		}
+		sizePerRack := int(cassDC.Spec.Size) / len(racks)
+		out[location] = getPodNames(Kluster.Name, cassDC.Name, i.Name, sizePerRack)
+	}
+	return out, nil
+}
+
+func getPodNames(clusterName string, DCName string, rackName string, rackSize int) []HostName {
+	out := []HostName{}
+	for i := 0; i < rackSize; i++ {
+		out = append(out, HostName(clusterName+"-"+DCName+"-"+rackName+"-sts-"+fmt.Sprint(i)))
+	}
+	return out
+}
+
+func cassDCFromKluster(Kluster k8ssandraapi.K8ssandraCluster, dcName string) (*cassdcapi.CassandraDatacenter, error) {
+	thisDC := k8ssandraapi.CassandraDatacenterTemplate{}
+	if len(Kluster.Spec.Cassandra.Datacenters) > 0 {
+		for _, i := range Kluster.Spec.Cassandra.Datacenters {
+			if i.Meta.Name == dcName {
+				thisDC = i
+			}
+		}
+	}
+	DCConfig := cassandrapkg.Coalesce(Kluster.Name, Kluster.Spec.Cassandra, &thisDC)
+	cassDC, err := cassandrapkg.NewDatacenter(
+		types.NamespacedName{
+			Namespace: Kluster.Namespace,
+			Name:      Kluster.Name,
+		},
+		DCConfig)
+	if err != nil {
+		return nil, err
+	}
+	return cassDC, nil
+}
+
+// GetHostMap gets the hostmap for a given CassandraBackup from IP sources to FQDN targets from the K8ssandraCluster and the backups returned by the Medusa gRPC client.
+// TODO: check for rack imbalances which may cause subtle errors here. Also need to check that source rack sizes are the same as destination rack sizes.
+func GetHostMap(Kluster k8ssandraapi.K8ssandraCluster, k8sbackup medusaapi.CassandraRestore, client backupGetter, ctx context.Context) (HostMappingSlice, error) {
+	sourceRacks, err := getSourceRacksIPs(k8sbackup, client, ctx)
+	if err != nil {
+		return nil, err
+	}
+	destRacks, err := getTargetRackFQDNs(Kluster, k8sbackup.Spec.CassandraDatacenter.Name)
+	if err != nil {
+		return nil, err
+	}
+	out := HostMappingSlice{}
+	for sourceRackLocation, sourceRackNodes := range sourceRacks {
```

Comment:
There are three solutions to this problem.

### Solution 1

In the current solution, rack names must match, and the racks must be the same size between the source and destination topologies, but they do **not** need to be balanced (i.e. all racks the same size). Additionally, users can control the mapping with precision.

### Solution 2

Under your proposal, rack names may differ, but the racks must be balanced and users cannot control the placements with precision.

### Solution 3

If we don't care about being able to control placement, we can eliminate requirements on both balance and naming via the following:
1. Add a function which takes a `map[nodeLocation]string` and returns `map[len(nodes)]nodeLocation`.
2. Put sources and targets though that function.
3. Do a "join" on the two maps via a for loop using len(nodes) as the key.

This has the advantage that racks may be unbalanced AND names need not match. But users still cannot control placement since (in the standard case where all racks are the same size) the source-destination mapping will be randomised.

In some ways, I'd rather force the rack names to match as it gives more control. But I'm not committed to any particular solution here.

Reply:
Discussion we had offline (with thanks to @adejanovski for the info):

* Network topology strategy actually allocates replicas based on rack names (replicas go to the racks alphabetically close to the primary replica).
* The EC2 aware snitch will actually modify rack names.

As a result of these two facts, it isn't appropriate to (a) force users to use the same rack names between source and destination clusters or (b) allow customers to change the token ordering relative to the rack-name alphabetical ordering.

Further changes to account for this logic incoming.

Reply:
>The EC2 aware snitch will actually modify rack names

Does it matter that we always use GossipingPropertyFileSnitch?

Reply:
I'm not sure @jsanda but it is a worthwhile callout. Based on conversation with Alexander, it sounded like he was also concerned about the replicas being allocated into the same racks (which happens based on the alphabetical ordering of their names). I didn't quite follow that part 100%.

@adejanovski your thoughts are welcome. I have to be honest, the most recent round of changes has left me less than happy with this code as I feel that the data structures etc. that I originally wrote (when I was going to force rack names to match) are now poorly suited to the new requirements around alphabetical ordering.

Reply:
> Does it matter that we always use GossipingPropertyFileSnitch?

No, it's fine. It's just worth mentioning the EC2Snitch because the source cluster could be using it.